test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/TestHubMessageEqualityComparer.cs

```diff
@@ -0,0 +1,107 @@
+            if (!string.Equals(x.InvocationId, y.InvocationId, StringComparison.Ordinal))
+            {
+                return false;
+            }
+
+            return InvocationMessagesEqual(x, y) || StreamItemMessagesEqual(x, y) || CompletionMessagesEqual(x, y);
+        }
+
+        private bool CompletionMessagesEqual(HubMessage x, HubMessage y)
+        {
+            return x is CompletionMessage left && y is CompletionMessage right &&
+                string.Equals(left.Error, right.Error, StringComparison.Ordinal) &&
+                left.HasResult == right.HasResult &&
+                (Equals(left.Result, right.Result) || SequenceEqual(left.Result, right.Result));
+        }
+
+        private bool StreamItemMessagesEqual(HubMessage x, HubMessage y)
+        {
+            return x is StreamItemMessage left && y is StreamItemMessage right &&
+                (Equals(left.Item, right.Item) || SequenceEqual(left.Item, right.Item));
+        }
+
+        private bool InvocationMessagesEqual(HubMessage x, HubMessage y)
+        {
+            return x is InvocationMessage left && y is InvocationMessage right &&
+                string.Equals(left.Target, right.Target, StringComparison.Ordinal) &&
+                ArgumentListsEqual(left.Arguments, right.Arguments) &&
+                left.NonBlocking == right.NonBlocking;
+        }
+
+        private bool ArgumentListsEqual(object[] left, object[] right)
+        {
+            if (left == right)
+            {
+                return true;
+            }
+
+            if (left == null || right == null || left.Length != right.Length)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < left.Length; i++)
+            {
+                if (!(Equals(left[i], right[i]) || SequenceEqual(left[i], right[i])))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private bool SequenceEqual(object left, object right)
+        {
+            if (left == null && right == null)
+            {
+                return true;
+            }
+
+            var leftEnumerable = left as IEnumerable;
+            var rightEnumerable = right as IEnumerable;
+            if (leftEnumerable == null || rightEnumerable == null)
+            {
+                return false;
+            }
+
+            var leftEnumerator = leftEnumerable.GetEnumerator();
+            var rightEnumerator = rightEnumerable.GetEnumerator();
+            var leftMoved = leftEnumerator.MoveNext();
+            var rightMoved = rightEnumerator.MoveNext();
+            for (; leftMoved && rightMoved; leftMoved = leftEnumerator.MoveNext(), rightMoved = rightEnumerator.MoveNext())
+            {
+                if (!Equals(leftEnumerator.Current, rightEnumerator.Current))
+                {
+                    return false;
+                }
+            }
+
+            return !leftMoved && !rightMoved;
```

Comment:
no, as per De Morgan law it would be !(A || B). I had it but I reverted since I find it cleaner.